### PR TITLE
EIP712 signer for ZkSync

### PR DESCRIPTION
**Pull request by mariopil**
This PR implements EIP712 signer support for ZkSync chain (related discussion #1229 ).

---

### Focus of this PR:
This PR focuses on versioning updates and code cleanup.

### Detailed summary:
- Updated version to '2.0.0-rc.3'
- Deleted unnecessary files and directories
- Updated target to 'ESNext' in tsconfig.json
- Added a pre.json file with initial versions for a release candidate (rc) tag
- Updated dependencies in package.json files
- Made changes to CSS styles in theme/index.css
- Made changes to chain configurations in various files
- Made changes to test files and type definitions

> The following files were skipped due to too many changes: `examples/clients_public-client/package.json`, `examples/clients_wallet-client/package.json`, `examples/blocks_fetching-blocks/package.json`, `examples/blocks_watching-blocks/package.json`, `examples/logs_block-event-logs/package.json`, `src/clients/createClient.ts`, `src/chains/definitions/ekta.ts`, `src/chains/definitions/spicy.ts`, `src/chains/opStack/types/chain.ts`, `src/utils/abi/encodeEventTopics.test.ts`, `examples/contracts_reading-contracts/package.json`, `src/chains/definitions/klaytnBaobab.ts`, `examples/contracts_writing-to-contracts/package.json`, `examples/contracts_deploying-contracts/package.json`, `examples/transactions_sending-transactions/package.json`, `src/types/misc.ts`, `examples/transactions_fetching-transactions/package.json`, `site/docs/clients/public.md`, `src/actions/public/getTransactionReceipt.ts`, `src/chains/definitions/ektaTestnet.ts`, `src/chains/definitions/eon.ts`, `src/chains/definitions/fuseSparknet.ts`, `src/actions/wallet/requestPermissions.ts`, `src/chains/definitions/oasys.ts`, `src/actions/wallet/watchAsset.ts`, `src/chains/definitions/vechain.ts`, `src/chains/definitions/polygonZkEvm.ts`, `src/chains/definitions/harmonyOne.ts`, `src/chains/zksync/chainConfig.ts`, `src/chains/definitions/taraxa.ts`, `src/actions/public/getBlock.ts`, `src/chains/definitions/hardhat.ts`, `src/chains/definitions/modeTestnet.ts`, `src/chains/definitions/plinga.ts`, `src/chains/utils/package.json`, `src/chains/definitions/shardeumSphinx.ts`, `src/chains/definitions/localhost.ts`, `src/chains/definitions/taraxaTestnet.ts`, `src/chains/definitions/opBNBTestnet.ts`, `src/chains/definitions/gobi.ts`, `src/chains/definitions/scrollSepolia.ts`, `examples/op-stack_deposit/index.html`, `src/chains/definitions/fuse.ts`, `src/chains/definitions/saigon.ts`, `src/chains/definitions/qMainnet.ts`, `src/chains/definitions/taikoJolnir.ts`, `src/chains/definitions/taikoTestnetSepolia.ts`, `src/chains/definitions/kroma.ts`, `src/utils/abi/encodeErrorResult.test.ts`, `src/chains/definitions/bxn.ts`, `src/chains/definitions/thunderTestnet.ts`, `src/chains/definitions/telos.ts`, `src/chains/definitions/classic.ts`, `src/utils/rpc.test.ts`, `src/chains/definitions/canto.ts`, `src/chains/definitions/qTestnet.ts`, `src/actions/getContract.bench.ts`, `src/chains/definitions/bearNetworkChainTestnet.ts`, `src/chains/definitions/ronin.ts`, `src/actions/wallet/requestAddresses.ts`, `src/chains/definitions/meterTestnet.ts`, `src/chains/definitions/telosTestnet.ts`, `src/actions/wallet/sendRawTransaction.ts`, `src/chains/definitions/foundry.ts`, `src/chains/definitions/haqqMainnet.ts`, `src/chains/definitions/cronosTestnet.ts`, `src/chains/definitions/metis.ts`, `src/chains/definitions/bearNetworkChainMainnet.ts`, `src/chains/definitions/neonDevnet.ts`, `src/utils/formatters/block.ts`, `src/chains/definitions/meter.ts`, `src/chains/definitions/evmosTestnet.ts`, `src/chains/definitions/zetachainAthensTestnet.ts`, `src/chains/definitions/holesky.ts`, `src/chains/definitions/mev.ts`, `src/chains/definitions/okc.ts`, `src/chains/zksync/formatters.test.ts`, `src/chains/definitions/flare.ts`, `src/chains/definitions/bronos.ts`, `src/utils/chain/defineChain.test.ts`, `site/.vitepress/config.ts`, `src/chains/definitions/kromaSepolia.ts`, `src/chains/definitions/cronos.ts`, `src/chains/definitions/moonbeamDev.ts`, `src/chains/definitions/xdc.ts`, `site/docs/clients/custom.md`, `src/chains/definitions/flareTestnet.ts`, `src/chains/definitions/haqqTestedge2.ts`, `src/chains/definitions/metisGoerli.ts`, `src/chains/definitions/songbirdTestnet.ts`, `.size-limit.json`, `src/chains/definitions/fibo.ts`, `src/chains/definitions/confluxESpace.ts`, `src/chains/definitions/songbird.ts`, `src/chains/definitions/evmos.ts`, `examples/op-stack_deposit/package.json`, `src/chains/definitions/mevTestnet.ts`, `src/chains/definitions/bxnTestnet.ts`, `src/chains/definitions/rolluxTestnet.ts`, `src/actions/test/sendUnsignedTransaction.test.ts`, `src/chains/definitions/nexi.ts`, `src/actions/wallet/switchChain.ts`, `src/chains/definitions/scrollTestnet.ts`, `src/chains/definitions/xdcTestnet.ts`, `src/chains/definitions/rollux.ts`, `src/chains/opStack/utils/getWithdrawalHashStorageSlot.test.ts`, `src/chains/definitions/crossbell.ts`, `src/chains/definitions/coreDao.ts`, `src/chains/definitions/gnosisChiado.ts`, `src/chains/definitions/bronosTestnet.ts`, `.github/workflows/verify.yml`, `src/utils/errors/getNodeError.ts`, `playgrounds/browser/package.json`, `src/chains/definitions/wanchainTestnet.ts`, `src/chains/definitions/bsc.ts`, `src/chains/definitions/opBNB.ts`, `src/chains/definitions/fantom.ts`, `src/chains/celo/formatters.ts`, `src/chains/definitions/arbitrumSepolia.ts`, `src/chains/definitions/boba.ts`, `src/chains/definitions/mantle.ts`, `src/chains/definitions/syscoin.ts`, `src/utils/abi/encodeFunctionResult.test.ts`, `src/chains/definitions/iotex.ts`, `.env.example`, `src/chains/definitions/klaytn.ts`, `src/chains/definitions/aurora.ts`, `src/chains/definitions/eos.ts`, `src/chains/definitions/filecoinCalibration.ts`, `src/chains/definitions/polygonZkEvmTestnet.ts`, `src/chains/definitions/pulsechain.ts`, `site/docs/clients/chains.md`, `src/chains/definitions/dogechain.ts`, `src/chains/definitions/nexilix.ts`, `src/chains/utils/index.test.ts`, `src/chains/definitions/avalanche.ts`, `src/chains/utils.ts`, `src/chains/definitions/syscoinTestnet.ts`, `src/chains/definitions/iotexTestnet.ts`, `src/chains/definitions/wanchain.ts`, `src/chains/definitions/fantomTestnet.ts`, `src/chains/definitions/auroraTestnet.ts`, `src/chains/opStack/actions/estimateInitiateWithdrawalGas.test.ts`, `src/chains/definitions/scroll.ts`, `src/chains/opStack/errors/withdrawal.ts`, `src/chains/definitions/mantleTestnet.ts`, `src/chains/definitions/edgeware.ts`, `examples/op-stack_deposit/tsconfig.json`, `src/chains/definitions/bscTestnet.ts`, `src/chains/definitions/dfk.ts`, `src/chains/definitions/avalancheFuji.ts`, `src/chains/definitions/gnosis.ts`, `.github/workflows/prune-tags.yml`, `site/docs/actions/wallet/prepareTransactionRequest.md`, `src/chains/definitions/pulsechainV4.ts`, `src/chains/definitions/eosTestnet.ts`, `src/chains/definitions/filecoinHyperspace.ts`, `src/chains/definitions/edgewareTestnet.ts`, `src/chains/opStack/types/withdrawal.ts`, `src/chains/opStack/actions/getL2Output.test.ts`, `src/chains/definitions/moonbeam.ts`, `src/accounts/types.ts`, `src/chains/opStack/types/contract.ts`, `src/utils/ccip.ts`, `src/actions/public/waitForTransactionReceipt.test.ts`, `src/chains/definitions/zhejiang.ts`, `src/chains/definitions/moonriver.ts`, `site/zksync/chains.md`, `src/chains/definitions/filecoin.ts`, `test/src/abis.ts`, `src/chains/definitions/arbitrumGoerli.ts`, `src/chains/definitions/moonbaseAlpha.ts`, `src/chains/definitions/polygon.ts`, `src/chains/opStack/chains.ts`, `src/chains/definitions/arbitrumNova.ts`, `src/actions/public/simulateContract.test.ts`, `src/chains/definitions/arbitrum.ts`, `src/clients/transports/webSocket.test.ts`, `src/actions/public/waitForTransactionReceipt.ts`, `.github/workflows/pull-request.yml`, `src/chains/definitions/skale/humanProtocol.ts`, `src/chains/definitions/goerli.ts`, `src/chains/definitions/sepolia.ts`, `src/chains/definitions/polygonMumbai.ts`, `src/chains/definitions/skale/exorde.ts`, `src/chains/definitions/skale/europa.ts`, `src/chains/opStack/decorators/walletL2.test.ts`, `src/chains/definitions/linea.ts`, `src/chains/definitions/skale/brawl.ts`, `src/chains/definitions/skale/nebula.ts`, `src/actions/public/estimateFeesPerGas.ts`, `src/chains/definitions/skale/titan.ts`, `src/chains/opStack/utils/getSourceHash.test.ts`, `src/chains/definitions/skale/razor.ts`, `src/chains/opStack/contracts.ts`, `src/chains/celo/index.ts`, `src/chains/definitions/skale/cryptoColosseum.ts`, `src/actions/wallet/addChain.ts`, `src/actions/wallet/sendTransaction.test.ts`, `src/chains/definitions/skale/calypso.ts`, `src/clients/createClient.test.ts`, `src/clients/createPublicClient.test.ts`, `src/types/transport.ts`, `src/chains/definitions/lineaTestnet.ts`, `src/chains/opStack/utils/getWithdrawals.ts`, `src/actions/public/estimateMaxPriorityFeePerGas.ts`, `src/types/account.ts`, `src/chains/definitions/skale/cryptoBlades.ts`, `site/op-stack/chains.md`, `src/chains/definitions/skale/nebulaTestnet.ts`, `test/setup.ts`, `src/chains/definitions/skale/europaTestnet.ts`, `src/chains/definitions/skale/chaosTestnet.ts`, `src/chains/definitions/skale/calypsoTestnet.ts`, `src/chains/definitions/skale/titanTestnet.ts`, `src/clients/createClient.test-d.ts`, `src/actions/public/getBlockNumber.ts`, `src/chains/opStack/utils/getL2TransactionHashes.ts`, `src/chains/opStack/actions/getTimeToProve.test.ts`, `src/types/eip1193.ts`, `src/chains/opStack/utils/extractTransactionDepositedLogs.ts`, `src/index.test.ts`, `src/types/rpc.ts`, `src/types/chain.test-d.ts`, `src/chains/opStack/utils/extractWithdrawalMessageLogs.ts`, `src/chains/opStack/utils/getWithdrawalHashStorageSlot.ts`, `src/chains/opStack/actions/estimateDepositTransactionGas.test.ts`, `src/utils/chain/defineChain.ts`, `.github/workflows/on-push-to-main.yml`, `src/chains/opStack/types/deposit.ts`, `src/chains/opStack/types/block.ts`, `src/utils/index.test.ts`, `src/actions/wallet/signTypedData.test-d.ts`, `src/chains/definitions/mainnet.ts`, `src/clients/createTestClient.test.ts`, `site/op-stack/utilities/fromOpaqueData.md`, `src/chains/index.ts`, `src/actions/public/watchContractEvent.test-d.ts`, `src/chains/opStack/utils/getL2TransactionHashes.test.ts`, `src/chains/celo/serializers.ts`, `src/chains/zksync/formatters.ts`, `site/op-stack/utilities/getWithdrawals.md`, `site/op-stack/utilities/getL2TransactionHashes.md`, `site/op-stack/utilities/getWithdrawalHashStorageSlot.md`, `site/docs/chains/serializers.md`, `src/chains/zksync/actions/sendTransaction.test.ts`, `src/types/utils.test-d.ts`, `src/utils/transaction/assertRequest.ts`, `src/chains/zksync/index.ts`, `src/chains/opStack/actions/waitToFinalize.test.ts`, `test/globalSetup.ts`, `test/src/zksync.ts`, `src/chains/opStack/utils/opaqueDataToDepositData.ts`, `src/chains/opStack/actions/waitForNextL2Output.test.ts`, `src/actions/public/estimateGas.test.ts`, `site/op-stack/utilities/extractTransactionDepositedLogs.md`, `src/chains/definitions/celoCannoli.ts`, `src/chains/opStack/utils/getL2TransactionHash.test.ts`, `site/op-stack/utilities/parseTransaction.md`, `src/clients/createWalletClient.test.ts`, `src/chains/definitions/zkSync.ts`, `src/chains/definitions/celo.ts`, `src/utils/abi/encodeDeployData.test-d.ts`, `site/docs/actions/public/verifyMessage.md`, `src/utils/signature/recoverTypedDataAddress.ts`, `src/utils/signature/verifyTypedData.ts`, `site/docs/chains/introduction.md`, `src/chains/opStack/actions/estimateContractL1Gas.test.ts`, `src/chains/opStack/actions/estimateContractL1Fee.test.ts`, `src/chains/opStack/utils/getL2TransactionHash.ts`, `src/chains/opStack/actions/estimateContractTotalFee.test.ts`, `src/chains/opStack/actions/estimateContractTotalGas.test.ts`, `src/chains/definitions/zora.ts`, `site/op-stack/utilities/extractWithdrawalMessageLogs.md`, `.github/CONTRIBUTING.md`, `src/chains/definitions/celoAlfajores.ts`, `src/actions/index.test.ts`, `src/chains/opStack/actions/getTimeToFinalize.test.ts`, `src/chains/definitions/zkSyncSepoliaTestnet.ts`, `src/chains/definitions/zkSyncTestnet.ts`, `src/utils/index.ts`, `package.json`, `src/chains/definitions/baseSepolia.ts`, `src/utils/typedData.ts`, `src/chains/definitions/zoraTestnet.ts`, `src/chains/definitions/pgn.ts`, `src/errors/encoding.ts`, `site/zksync/client.md`, `src/accounts/utils/signTypedData.ts`, `src/actions/public/verifyTypedData.ts`, `src/chains/definitions/optimismSepolia.ts`, `src/chains/definitions/pgnTestnet.ts`, `src/chains/optimism/formatters.test-d.ts`, `src/chains/definitions/baseGoerli.ts`, `src/chains/zksync/actions/signTransaction.test.ts`, `src/actions/public/watchBlockNumber.ts`, `src/chains/definitions/zoraSepolia.ts`, `site/zksync.md`, `src/clients/transports/fallback.ts`, `src/chains/definitions/optimism.ts`, `src/chains/definitions/base.ts`, `src/actions/wallet/signTransaction.ts`, `src/chains/opStack/utils/getSourceHash.ts`, `src/chains/opStack/utils/getWithdrawals.test.ts`, `src/chains/opStack/actions/estimateFinalizeWithdrawalGas.test.ts`, `src/actions/public/watchPendingTransactions.ts`, `src/chains/definitions/optimismGoerli.ts`, `site/docs/actions/public/verifyTypedData.md`, `src/actions/wallet/sendTransaction.ts`, `src/utils/formatters/transaction.ts`, `site/docs/chains/zksync.md`, `site/op-stack/utilities/getL2TransactionHash.md`, `src/utils/signature/hashTypedData.ts`, `src/utils/abi/decodeDeployData.test-d.ts`, `src/chains/celo/parsers.ts`, `src/chains/opStack/actions/estimateL1Fee.test.ts`, `src/chains/opStack/actions/estimateL1Gas.test.ts`, `src/utils/abi/encodeErrorResult.test-d.ts`, `src/chains/opStack/actions/estimateTotalFee.test.ts`, `test/src/opStack.ts`, `src/chains/opStack/actions/estimateTotalGas.test.ts`, `src/actions/getContract.test.ts`, `src/actions/wallet/deployContract.test-d.ts`, `src/utils/abi/decodeEventLog.test-d.ts`, `src/chains/opStack/utils/extractWithdrawalMessageLogs.test.ts`, `src/chains/opStack/actions/getTimeToNextL2Output.test.ts`, `src/actions/public/estimateGas.ts`, `src/actions/public/getFilterLogs.ts`, `src/chains/opStack/utils/opaqueDataToDepositData.test.ts`, `src/utils/abi/decodeDeployData.ts`, `src/chains/opStack/actions/waitToProve.test.ts`, `site/op-stack/actions/getL2Output.md`, `src/chains/opStack/actions/waitToFinalize.ts`, `site/op-stack/client.md`, `site/docs/chains/celo.md`, `site/op-stack/utilities/getSourceHash.md`, `src/actions/public/getLogs.ts`, `site/op-stack.md`, `src/actions/wallet/writeContract.test-d.ts`, `src/actions/public/watchBlocks.ts`, `src/utils/abi/decodeFunctionData.test-d.ts`, `src/chains/zksync/serializers.ts`, `src/utils/abi/encodeDeployData.ts`, `src/chains/zksync/eip712signers.ts`, `src/utils/chain/extractChain.test.ts`, `src/actions/public/getFilterChanges.ts`, `site/op-stack/actions/waitToProve.md`, `src/chains/optimism/formatters.ts`, `site/op-stack/utilities/serializeTransaction.md`, `src/utils/abi/encodeEventTopics.test-d.ts`, `src/chains/opStack/actions/estimateTotalGas.ts`, `src/utils/abi/decodeFunctionData.ts`, `src/actions/wallet/signTypedData.ts`, `src/chains/zksync/decorators/eip712.ts`, `src/chains/opStack/serializers.ts`, `src/actions/public/readContract.test-d.ts`, `src/chains/opStack/types/transaction.ts`, `site/op-stack/actions/waitForNextL2Output.md`, `src/chains/opStack/actions/getTimeToProve.ts`, `src/actions/index.ts`, `src/chains/opStack/actions/getL2Output.ts`, `src/chains/opStack/actions/estimateTotalFee.ts`, `src/utils/buildRequest.ts`, `src/clients/transports/fallback.test.ts`, `src/chains/opStack/parsers.ts`, `src/utils/abi/encodeFunctionResult.test-d.ts`, `src/utils/abi/encodeFunctionData.test-d.ts`, `site/docs/chains/fees.md`, `src/actions/wallet/deployContract.ts`, `site/op-stack/actions/getTimeToProve.md`, `site/op-stack/actions/getTimeToNextL2Output.md`, `site/op-stack/actions/getWithdrawalStatus.md`, `examples/op-stack_deposit/index.tsx`, `src/chains/opStack/actions/waitToProve.ts`, `src/chains/zksync/decorators/eip712.test.ts`, `src/clients/decorators/wallet.ts`, `site/op-stack/actions/waitToFinalize.md`, `src/chains/opStack/parsers.test.ts`, `src/actions/public/estimateContractGas.ts`, `src/chains/zksync/types.ts`, `src/utils/abi/parseEventLogs.ts`, `src/chains/opStack/actions/waitForNextL2Output.ts`, `src/utils/abi/decodeErrorResult.ts`, `src/actions/public/watchEvent.ts`, `src/chains/opStack/actions/buildInitiateWithdrawal.test.ts`, `src/chains/opStack/actions/estimateFinalizeWithdrawalGas.ts`, `src/utils/abi/encodeFunctionResult.ts`, `site/op-stack/actions/getTimeToFinalize.md`, `src/chains/opStack/decorators/walletL2.ts`, `src/chains/opStack/actions/estimateInitiateWithdrawalGas.ts`, `src/utils/abi/encodeErrorResult.ts`, `src/actions/public/readContract.ts`, `src/utils/abi/encodeEventTopics.ts`, `src/chains/celo/formatters.test-d.ts`, `src/package.json`, `src/chains/zksync/eip712signers.test.ts`, `src/chains/opStack/actions/finalizeWithdrawal.ts`, `src/actions/public/simulateContract.test-d.ts`, `src/chains/opStack/actions/estimateProveWithdrawalGas.ts`, `src/chains/opStack/actions/getWithdrawalStatus.test.ts`, `src/types/typedData.ts`, `src/utils/abi/encodeFunctionData.ts`, `src/chains/opStack/decorators/publicL2.test.ts`, `src/chains/zksync/actions/signTransaction.ts`, `src/chains/opStack/actions/estimateDepositTransactionGas.ts`, `src/chains/opStack/actions/getTimeToFinalize.ts`, `src/chains/opStack/actions/buildInitiateWithdrawal.ts`, `src/chains/opStack/actions/estimateContractL1Fee.ts`, `src/utils/abi/getAbiItem.test-d.ts`, `src/actions/public/multicall.ts`, `src/chains/opStack/actions/estimateContractL1Gas.ts`, `src/chains/opStack/actions/estimateContractTotalFee.ts`, `src/chains/opStack/actions/proveWithdrawal.ts`, `src/chains/opStack/actions/estimateL1Fee.ts`, `src/chains/opStack/actions/estimateContractTotalGas.ts`, `src/chains/opStack/actions/estimateL1Gas.ts`, `site/docs/contract/getContract.md`, `site/op-stack/actions/estimateL1Fee.md`, `site/op-stack/actions/estimateTotalGas.md`, `site/op-stack/actions/estimateTotalFee.md`, `site/op-stack/actions/estimateL1Gas.md`, `src/chains/opStack/actions/buildProveWithdrawal.ts`, `src/utils/abi/decodeFunctionResult.test-d.ts`, `src/actions/public/getContractEvents.ts`, `site/docs/chains/formatters.md`, `site/op-stack/actions/buildProveWithdrawal.md`, `site/docs/migration-guide.md`, `src/chains/optimism/formatters.test.ts`, `src/chains/zksync/actions/sendTransaction.ts`, `src/chains/zksync/formatters.test-d.ts`, `src/chains/opStack/actions/buildDepositTransaction.ts`, `src/actions/wallet/writeContract.ts`, `src/actions/public/createContractEventFilter.ts`, `src/chains/opStack/actions/initiateWithdrawal.ts`, `src/chains/opStack/actions/getWithdrawalStatus.ts`, `src/chains/opStack/actions/getTimeToNextL2Output.ts`, `src/chains/opStack/serializers.test.ts`, `src/utils/abi/getAbiItem.ts`, `site/op-stack/actions/estimateInitiateWithdrawalGas.md`, `site/op-stack/actions/buildInitiateWithdrawal.md`, `src/utils/buildRequest.test.ts`, `src/chains/zksync/actions/prepareTransactionRequest.ts`, `src/utils/abi/decodeEventLog.ts`, `src/chains/opStack/actions/depositTransaction.ts`, `site/.vitepress/sidebar.ts`, `site/op-stack/actions/estimateContractL1Fee.md`, `site/op-stack/actions/estimateContractL1Gas.md`, `site/op-stack/actions/estimateFinalizeWithdrawalGas.md`, `site/op-stack/actions/estimateContractTotalFee.md`, `site/op-stack/actions/estimateContractTotalGas.md`, `src/actions/public/multicall.test-d.ts`, `site/op-stack/actions/buildDepositTransaction.md`, `src/chains/opStack/decorators/walletL1.ts`, `src/actions/wallet/prepareTransactionRequest.test.ts`, `src/utils/abi/parseEventLogs.test.ts`, `src/utils/abi/decodeFunctionResult.ts`, `src/chains/opStack/actions/estimateProveWithdrawalGas.test.ts`, `src/chains/opStack/decorators/walletL1.test.ts`, `src/chains/celo/parsers.test.ts`, `src/actions/public/multicall.test.ts`, `src/chains/opStack/index.ts`, `src/actions/public/watchContractEvent.ts`, `src/chains/opStack/actions/buildProveWithdrawal.test.ts`, `src/actions/getContract.test-d.ts`, `site/zksync/actions/signTransaction.md`, `src/clients/decorators/public.ts`, `site/op-stack/actions/estimateDepositTransactionGas.md`, `site/op-stack/actions/estimateProveWithdrawalGas.md`, `site/zksync/actions/sendTransaction.md`, `site/op-stack/actions/finalizeWithdrawal.md`, `src/types/utils.ts`, `src/chains/opStack/actions/initiateWithdrawal.test.ts`, `site/zksync/actions/prepareTransactionRequest.md`, `src/actions/public/simulateContract.ts`, `src/types/multicall.ts`, `src/actions/wallet/prepareTransactionRequest.ts`, `src/chains/opStack/actions/finalizeWithdrawal.test.ts`, `site/docs/contract/parseEventLogs.md`, `src/types/chain.ts`, `src/chains/opStack/decorators/publicL1.test.ts`, `site/op-stack/actions/initiateWithdrawal.md`, `src/index.ts`, `src/chains/opStack/actions/depositTransaction.test.ts`, `site/op-stack/actions/depositTransaction.md`, `site/op-stack/actions/proveWithdrawal.md`, `src/types/contract.test-d.ts`, `src/chains/zksync/serializers.test.ts`, `src/constants/abis.ts`, `src/constants/abis.test.ts`, `site/op-stack/guides/deposits.md`, `src/chains/opStack/decorators/publicL2.ts`, `src/chains/opStack/abis.ts`, `site/op-stack/guides/withdrawals.md`, `src/chains/opStack/decorators/publicL1.ts`, `src/types/contract.ts`, `src/chains/celo/serializers.test.ts`, `src/chains/opStack/actions/proveWithdrawal.test.ts`, `src/actions/getContract.ts`, `src/chains/opStack/actions/buildDepositTransaction.test.ts`, `src/chains/opStack/utils/extractTransactionDepositedLogs.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by mariopil** [comment]
There are still some type errors for which I would require a guide to fix them as I'm struggling a bit with the type system here. There's also one added test failing which I'm still working on and documentation is missing (work also ongoing).

**Comment by mariopil** [comment]
I have updated tests and docs. There is one issue with the tests though - EIP712 transactions will not work with anvil & forked ZkSync testnet node, as anvil is an Ethereum node. Currently in the tests related to EIP712 transactions I've set the http transport url to be directly ZkSync testnet, however this might cause some problems if testnet accounts will run out of tokens. Local ZkSync node (docker image) could be used but that would require additional setup for tests.

File: src/chains/zksync/utils/defineChain.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    eip712domain = chain.eip712domain,
+  } = config
+  return {
+    ...chain,
+    fees,
+    formatters,
+    serializers,
+    eip712domain,
+  } as unknown as Assign<chain, ChainConfigEIP712<formatters>>
+}
```
Comment (jxom) [inline src/chains/zksync/utils/defineChain.ts:25]: I'd be okay with amending the `ChainConfig` type in Viem to have a `custom` property that allows arbitrary chain data so we don't have to go through the ceremony of forking `defineChain`.

```diff
+ type Custom = { [key: string]: unknown }

export type ChainConfig<
  formatters extends ChainFormatters | undefined = ChainFormatters | undefined,
+ custom extends Custom | undefined = Custom | undefined.  
> = {
+ custom?: custom | undefined
  /**
   * Modifies how chain data structures (ie. Blocks, Transactions, etc)
   * are formatted & typed.
   */
  formatters?: formatters | undefined
  /** Modifies how data (ie. Transactions) is serialized. */
  serializers?: ChainSerializers<formatters> | undefined
  /** Modifies how fees are derived. */
  fees?: ChainFees<formatters> | undefined
}
```

Then in usage, we can do `client.chain?.custom?.eip712domain`

File: src/chains/zksync/types/transaction.ts
Comment (jxom) [inline src/chains/zksync/types/transaction.ts:1]: Can we put these types in `zksync/types.ts`?

File: src/chains/zksync/types/eip712signer.ts
Comment (jxom) [inline src/chains/zksync/types/eip712signer.ts:1]: Can we put these types in `zksync/types.ts`?

**Review by jxom — COMMENTED** [review]
It's looking good! Left a few comments. A few high level things:

- Would we be able to branch this PR off the `beta` branch (v2)? We are planning on releasing v2 in a couple of weeks, and it will mean we won't have to go through the hassle of rebasing these changes into v2. There shouldn't be any breaking changes in v2 that conflict with the code here, so I anticipate branching off `beta` shouldn't be too bad. The only breaking change that will conflict is the change in the `defineChain` API.
- Can we make sure that the type errors are fixed (`bun run typecheck`)?
- Can we make sure the tests pass, and that we have full test coverage over modified files (can check with `bun run test:cov zksync`)? Tests are failing when I try to run them.

**Comment by roberts-ivanovs** [comment]
Hey, Roberts from Eiger here (colleague of @mariopil). Thanks for the feedback!
To let you know, we started working on fixing all the suggestions you mentioned, currently there's a [draft PR in our forked repo here](https://github.com/eigerco/viem/pull/2) (already rebased onto `beta`). Post-rebase fixing all the type issues has been a challenge 😄 
We'll update this PR (or create a new one and close this - feel free to voice your preference) once we get to a more stable version.

Happy Holidays! 🎄

**Comment by tmm** [comment]
Sounds good. We moved to the `rc` branch yesterday if you want to pull in those changes (there aren't many so far).

**Comment by jxom** [comment]
Closing in favour of #1665.

**Pull request closed without merging**